### ISSUE TITLE
Help test version 2.5.0.dev

### ISSUE THREAD

**Issue by jotyGill** [jotyGill/openpyn-nordvpn#115]
Hi all, After being busy with other things, I am back to improving openpyn. 

There have been some improvements/changes since the current version 2.4.2 #116, notably the installation path of non python files (configs and .sh scripts) is now same as the installation path for the python files 2ceead7df5624f1735189576a651cf1a0d09df2b (thanks to 
@xcambar) 263b4e78c5c490e900ec50af0c88f14025d5b799. Others include 
Error handling when 'gi' and 'Notify' not found (non gui os) : no need for --silent
Speedup "--daemon" mode time 
Fallback options for older versions of unzip and wget.
Proper killing of management-client with -k 

**Clould you please try the dev release** so we know that the version works on different OS's.
Specially MacOS and ASUSWRT-Merlin (@1951FDG).

Remove openpyn with
`sudo pip3 uninstall openpyn`
Also:
On Fedora 
`sudo rm -r /usr/share/openpyn /usr/lib/python3.*/site-packages/openpyn* /usr/local/lib/python3.*/site-packages/openpyn* /usr/local/bin/openpyn`

On Ubuntu
`sudo rm -r /usr/local/lib/python3.5/dist-packages/openpyn /usr/local/bin/openpyn /usr/local/bin/openpyn`

Install using pip
`sudo python3 -m pip install openpyn==2.5.0.dev5`

Or by cloning test branch
`git clone --branch test https://github.com/jotyGill/openpyn-nordvpn.git && cd openpyn-nordvpn`
`sudo python3 setup.py install`

Thank you.

**Comment by 1951FDG** [comment]
Hi,

Just finished testing on MacOS, 

One thing I noticed was:

Credentials is saved in `/usr/local/lib/python3.6/site-packages/credentials` and ovpn_tcp and ovpn_udp is saved in `/usr/local/lib/python3.6/site-packages/files`, is this the desired behaviour ? I thought they were supposed to be in: `/usr/local/lib/python3.6/site-packages/openpyn` ?

Commit b9d2903 introduced an error related to ping, ping output produced on MacOS is `b'4 packets transmitted, 4 packets received, 0.0% packet loss\n`

I also made two commits in my test branch to address this issue, https://github.com/1951FDG/openpyn-nordvpn/tree/test


I can do a pull request to your test from my test branch, is that ok?

I also want to pull in some commits from my other branch, into your test, such as changes to readme, and updates to ASUSWRT-Merlin related files, for example to support openvpn_options which was added lately,  is that ok?

I still have to test compatibility on ASUSWRT-Merlin. Waiting on input before continuing...

**Comment by jotyGill** [comment]
Absolutely my man, I really appreciate the effort you are putting into this project :) 
I see, yeah these folders should be installed in /openpyn. It's a simple fix, I'll do that after receiving your commits.

**Comment by jotyGill** [comment]
ok, it isn't what I thought I was. In linux it seems to be working as expected. Can you try out dev2 and run
`openpyn -v` # should print path locations
and 
`sudo openpyn --init`
check the locations

**Comment by yuriw** [comment]
Ok installed on ubuntu 16.04
Had to do some extra installation, e.g. `sudo apt-get install python3-setuptools`
We need to make sure that in final release all dependencies are listed in one line

Generally everything looks working fine.

The only exception I see is `-f` option.  It does not seem to resolve hosts' names after connecting.

steps:

without `-f`
`sudo openpyn us`
`curl ipinfo.io/ip
209.58.139.260` ##check external IP 

with `-f`
`sudo openpyn us -f`
`curl ipinfo.io/ip
curl: (6) Could not resolve host: ipinfo.io`

Another issue I noticed, after first time killing `sudo openpyn us` process by Ctrl-C the script automaticaly resrtarts, on the second try it works.

Hope it helps :)

**Comment by yuriw** [comment]
I also noticed that after leaving `openpyn` for long time, 24+ h, the system became tired and very slow, CPU running at ~100%, trying to verify VPN

**Comment by 1951FDG** [comment]
Hi, same issue!

install_dir /usr/local/lib/python3.6/site-packages/openpyn
__basefilepath__ /usr/local/lib/python3.6/site-packages/
credentials file  /usr/local/lib/python3.6/site-packages/credentials

**Comment by 1951FDG** [comment]
Shouldn't it be:

`credentials_file_path = install_dir + "/" + "credentials"`

instead of

`credentials_file_path = __basefilepath__ + "credentials"`

or maybe let me ask this question instead,

why is this being done to determine `__basefilepath__` ?

`__basefilepath__ = install_dir[:install_dir.rfind('/', 0, -1) + 1]`

I hope I'm not confused here, but isn't the goal for `__basefilepath__` to be:

`__basefilepath__ = /usr/local/lib/python3.6/site-packages/openpyn/`

`__basefilepath__ = os.path.dirname(os.path.abspath(__file__)) + "/"`

**Comment by jotyGill** [comment]
@1951FDG 

Ok, I am using Fedora and python packages here get installed like this
`/usr/local/lib/python3.6/site-packages/openpyn-2.5.0.dev2-py3.6.egg/openpyn`
The older setup was working fine for me. I was just keeping the other files seperate from the .py files in "/openpyn". I see not all distro work the same in this manner, so I have taken your suggestion to keep these files in "/openpyn".

Now in Fedora they get installed in
`/usr/local/lib/python3.6/site-packages/openpyn-2.5.0.dev2-py3.6.egg/openpyn/*`
On Ubuntu 
`/usr/local/lib/python3.5/dist-packages/openpyn/*`
So all is good :)

**Comment by yuriw** [comment]
@jotyGill I tried several time and still see -f no resolve issue.
I use -x every time because otherwise after disconnect system won't resolve either.

What I noticed that every with -f connection is made to us579 (why the same server?), see log tail:

Mon Mar 26 07:49:45 2018 ++ Certificate has EKU (str) TLS Web Server Authentication, expects TLS Web Server Authentication
Mon Mar 26 07:49:45 2018 VERIFY EKU OK
Mon Mar 26 07:49:45 2018 VERIFY OK: depth=0, C=PA, ST=PA, L=Panama, O=NordVPN, OU=NordVPN, CN=us579.nordvpn.com, name=NordVPN, emailAddress=cert@nordvpn.com
Mon Mar 26 07:49:45 2018 Data Channel Encrypt: Cipher 'AES-256-CBC' initialized with 256 bit key
Mon Mar 26 07:49:45 2018 Data Channel Encrypt: Using 512 bit message hash 'SHA512' for HMAC authentication
Mon Mar 26 07:49:45 2018 Data Channel Decrypt: Cipher 'AES-256-CBC' initialized with 256 bit key
Mon Mar 26 07:49:45 2018 Data Channel Decrypt: Using 512 bit message hash 'SHA512' for HMAC authentication
Mon Mar 26 07:49:45 2018 Control Channel: TLSv1.2, cipher TLSv1/SSLv3 ECDHE-RSA-AES256-GCM-SHA384, 2048 bit RSA
Mon Mar 26 07:49:45 2018 [us579.nordvpn.com] Peer Connection Initiated with [AF_INET]209.58.139.48:1194
install_dir /usr/local/lib/python3.5/dist-packages/openpyn-2.5.0.dev2-py3.5.egg/openpyn
__basefilepath__ /usr/local/lib/python3.5/dist-packages/openpyn-2.5.0.dev2-py3.5.egg/
Mon Mar 26 07:49:46 2018 MANAGEMENT: Client connected from [AF_INET]127.0.0.1:7015
Mon Mar 26 07:49:47 2018 SENT CONTROL [us579.nordvpn.com]: 'PUSH_REQUEST' (status=1)
Mon Mar 26 07:49:47 2018 PUSH: Received control message: 'PUSH_REPLY,redirect-gateway def1,sndbuf 524288,rcvbuf 524288,dhcp-option DNS 103.86.96.100,dhcp-option DNS 103.86.99.100,route-gateway 10.8.8.1,topology subnet,ping 60,ping-restart 180,ifconfig 10.8.8.25 255.255.255.0,peer-id 14'
Mon Mar 26 07:49:47 2018 OPTIONS IMPORT: timers and/or timeouts modified
Mon Mar 26 07:49:47 2018 OPTIONS IMPORT: --sndbuf/--rcvbuf options modified
Mon Mar 26 07:49:47 2018 Socket Buffers: R=[212992->425984] S=[212992->425984]
Mon Mar 26 07:49:47 2018 OPTIONS IMPORT: --ifconfig/up options modified
Mon Mar 26 07:49:47 2018 OPTIONS IMPORT: route options modified
Mon Mar 26 07:49:47 2018 OPTIONS IMPORT: route-related options modified
Mon Mar 26 07:49:47 2018 OPTIONS IMPORT: --ip-win32 and/or --dhcp-option options modified
Mon Mar 26 07:49:47 2018 OPTIONS IMPORT: peer-id set
Mon Mar 26 07:49:47 2018 OPTIONS IMPORT: adjusting link_mtu to 1637
Mon Mar 26 07:49:47 2018 ROUTE_GATEWAY 192.168.90.1/255.255.255.0 IFACE=ens3 HWADDR=52:54:00:41:9e:4a
Mon Mar 26 07:49:47 2018 TUN/TAP device tun0 opened
Mon Mar 26 07:49:47 2018 TUN/TAP TX queue length set to 100
Mon Mar 26 07:49:47 2018 do_ifconfig, tt->ipv6=0, tt->did_ifconfig_ipv6_setup=0
Mon Mar 26 07:49:47 2018 /sbin/ip link set dev tun0 up mtu 1500
Mon Mar 26 07:49:47 2018 /sbin/ip addr add dev tun0 10.8.8.25/24 broadcast 10.8.8.255
Mon Mar 26 07:49:47 2018 /usr/local/lib/python3.5/dist-packages/openpyn-2.5.0.dev2-py3.5.egg/openpyn/scripts/update-resolv-conf.sh tun0 1500 1637 10.8.8.25 255.255.255.0 init
dhcp-option DNS 162.242.211.137
dhcp-option DNS 78.46.223.24
dhcp-option DNS 208.67.222.220
/etc/resolvconf/update.d/libc: Warning: /etc/resolv.conf is not a symbolic link to /run/resolvconf/resolv.conf
Mon Mar 26 07:49:47 2018 /sbin/ip route add 209.58.139.48/32 via 192.168.90.1
Mon Mar 26 07:49:47 2018 /sbin/ip route add 0.0.0.0/1 via 10.8.8.1
Mon Mar 26 07:49:47 2018 /sbin/ip route add 128.0.0.0/1 via 10.8.8.1
Mon Mar 26 07:49:47 2018 Initialization Sequence Completed

and after no resolution `curl ipinfo.io/ip
curl: (6) Could not resolve host: ipinfo.io`

**Comment by jotyGill** [comment]
@yuriw 
I am still unable to reproduce the name resolve error (used Fedora 27, Ubuntu 16.04). Likely /etc/resolv.conf isn't getting updated or accepted on your PC.
What is the output of `sudo cat /etc/resolv.conf` after establishing a connection using "-f"?

**Comment by yuriw** [comment]
@jotyGill 

before starting `openpy` 
$ cat /etc/resolv.conf nameserver 127.0.1.1
search wawona

after starting `openpy us`
$ cat /etc/resolv.conf 
nameserver 127.0.1.1
search wawona
$ curl ipinfo.io/ip
170.178.182.12

after starting `openpy us -f`
$ cat /etc/resolv.conf 
nameserver 127.0.1.1
$ curl ipinfo.io/ip
curl: (6) Could not resolve host: ipinfo.io

stopped and : 
$ openpyn -x
install_dir /usr/local/lib/python3.5/dist-packages/openpyn-2.5.0.dev2-py3.5.egg/openpyn
__basefilepath__ /usr/local/lib/python3.5/dist-packages/openpyn-2.5.0.dev2-py3.5.egg/
credentials file  /usr/local/lib/python3.5/dist-packages/openpyn-2.5.0.dev2-py3.5.egg/credentials
Flushing iptables INPUT and OUTPUT chains AND Applying default Rules
Terminated

restored all to normal

**Comment by yuriw** [comment]
@jotyGill On Note 2 from the about comment:  I can easily reproduce it and it is trigged by `-f` arg, after running `openpy us -f` Plex and Emby running on the same ubuntu box become inaccessible 

Any ideas why ?

**Comment by jotyGill** [comment]
As I suspected, it's not an issue with "-f" option but with /etc/resolv.conf not getting updated. name resolution is not working because "-f" is doing its job of blocking all the traffic except through the tunnel. "nameserver 127.0.1.1" should have been replaced by Nord's dns servers, so the dns requests can go through the tunnel.

please reboot when testing one approach after the other. For approach 2 and 3 run using "--skip-dns-patch" 

First of all, replace the currently used "scripts/" file with "https://github.com/masterkorp/openvpn-update-resolv-conf/blob/master/update-resolv-conf.sh", make it executable and run openpyn.

Second approach: edit /etc/resolvconf/resolv.conf.d/base 
add
`nameserver 162.242.211.137`
`nameserver 78.46.223.24`

and run `sudo resolvconf -u` and test

3rd: run this [manual-dns-patch.zip](https://github.com/jotyGill/openpyn-nordvpn/files/1858228/manual-dns-patch.zip) script and test.

__________________________________________________________________________
You may have udhcp causing rewrite of resolv.conf . https://unix.stackexchange.com/questions/174349/what-overwrites-etc-resolv-conf-on-every-boot#174356 .

**Comment by yuriw** [comment]
@jotyGill ack, I won't be able to test this in next day or two, any chance all this will be pushed to the test branch by then?

**Comment by yuriw** [comment]
@jotyGill 
1st approach steps:

changed the `update-resolv-conf.sh` script
run `openpyn us -f --skip-dns-patch` => name resolution works fine
ctrl-c and run `openpyn -x` => name resolution works fine

**Comment by yuriw** [comment]
@jotyGill 
FYI I did confirmed that ports, when applications like plex, emby and x11vnc won't respond (aka "with /etc/resolv.conf not getting updated") is caused by the script.  I tried directly with `openvpn` and all works as expected.  

It's very annoying.

**Comment by jotyGill** [comment]
I have been busy with other things. The patch for this issue on newer Ubuntu OS's, was developed by karlicoss. I just haven't been able to test it and push it through. I also have some commits that need to be merged. (locations API changed). Hopefully I will get to work on it soon.
**In the mean time use "--skip-dns-patch".**

**Comment by yuriw** [comment]
@jotyGill I see issue related to `after running openpy us -f Plex and Emby running on the same ubuntu box become inaccessible` is still in openpyn 2.6.0
It prevent me us it on my ubuntu box, still can use on VM :)

**Comment by jotyGill** [comment]
closing old thread.
